### ISSUE TITLE
bug: Hydrate: serializeShadowRoot: false does not render content of the shadow dom of components.

### ISSUE THREAD

**Issue by mayerraphael** [ionic-team/stencil#5924]
### Prerequisites

- [X] I have read the [Contributing Guidelines](https://github.com/ionic-team/stencil/blob/main/CONTRIBUTING.md).
- [X] I agree to follow the [Code of Conduct](https://github.com/ionic-team/stencil/blob/main/CODE_OF_CONDUCT.md).
- [X] I have searched for [existing issues](https://github.com/ionic-team/stencil/issues) that already report this problem, without success.

### Stencil Version

4.20.0

### Current Behavior

When using `serializeShadowRoot: false`, the contents of the components are incorrectly rendered. Slot projection seems completly broken.

### With Stencil 4.19+

For some components, the full render() method content is missing, only the slotted object is there with no slot:

![image](https://github.com/user-attachments/assets/c56cba95-5c4a-4300-8187-5db27767d0f3)

### Same component with Stencil 4.18.3

![image](https://github.com/user-attachments/assets/b841164b-aa9c-46c2-a7f4-100d1ddd2c7e)

There are also another problem with the `renderToString` options.

```ts
const { html } = await stencil.renderToString(appHtml, {  }); 
```
This renders like setting `serializeShadowRoot: true` explicitly even though default is false. But null seems not to be equal to false here. The option is nullable.

![image](https://github.com/user-attachments/assets/8bbac07b-c5a1-4067-9207-2a839fe967b1)

serializeSHadowRoot must be explicitly set to false to use the old, but broken, behavior.

### Expected Behavior

The old hydration concept to work fine again. Tested against 4.13.0 and 4.18.3, which works fine.

### System Info

```shell
Stencil 4.20.0 with React Output Target 0.5.3. (latest stable).
```

### Steps to Reproduce

Create a component with a (default) slot:

```tsx
render() {
    return (
      <Host>
        <div>
          <slot></slot>
        </div>
      </Host>
    );
  }
```

Render the component with slot content:

```jsx
<MyWhateverComponent>Abc</MyWhateverComponent>
```

The generate html is broken:

```html
<my-whatever-component class="sc-my-whatever-component-h hydrated" s-id="1">
  <!--r.1-->
  Abc
  <div class="sc-my-whatever-component" c-id="1.0.0.0">
     <slot class="sc-my-whatever-component" c-id="1.1.1.0"></slot>
  </div>
</my-whatever-component>
```

See `nextjs-app-pagerouter` example in my repository: https://github.com/mayerraphael/stencil-dsd-ssr-playground

### Code Reproduction URL

https://github.com/mayerraphael/stencil-dsd-ssr-playground

**Comment by christian-bromann** [comment]
Thanks for reporting the issue. To clarify the issue: given a shadow component (with `shadow: true` flag set), if I render the component to string, e.g. 

```js
const { html } = await renderToString(`<my-whatever-component>Abc</my-whatever-component>`, {
  fullDocument: false,
  prettyHtml: true,
})

console.log(html);
```

I expect the slot to be replaced with the actual light dom, e.g.

```
    <my-whatever-component class="hydrated sc-my-whatever-component-h" s-id="1">
      <!--r.1-->
      <!--o.0.1.-->
      <div c-id="1.0.0.0" class="sc-my-whatever-component sc-my-whatever-component-s">
        <!--s.1.1.1.0.-->
        <!--t.0.1-->
        Abc
      </div>
    </my-whatever-component>
```

however the last stable of version of Stencil returns:

```
    <my-whatever-component class="hydrated sc-my-whatever-component-h" s-id="1">
      <!--r.1-->
      Abc
      <div c-id="1.0.0.0" class="sc-my-whatever-component">
        <slot c-id="1.1.1.0" class="sc-my-whatever-component"></slot>
      </div>
    </my-whatever-component>
```

**Comment by christian-bromann** [comment]
Thanks for reporting the issue and apology that we didn't clearly communicated this change. So let me do this here:

As we've further improved support for declarative Shadow DOM, we've realised that it wasn't feasible for us to allow users to render a `shadow` component as `scoped` component after the component has been compiled. The reason is that Stencil compiles the styles for either rendering a component is `shadow` or `scoped` mode. As part of the compilation step Stencil puts these compiled styles into the hydrate module at which point they can't be transformed anymore to support the other mode. Knowing that this would introduce a breaking change for this particular feature, we've decided as a team to move forward with it and introduce the change as part of `v4.19.0` (see PR #5892).

Our recommendation moving forward is to serialize all components marked with `shadow: true` as a declarative shadow DOM.

It would be super helpful if you could explain your use case and reasoning why it is beneficial for you to serialize shadow components as scoped components. Supporting both modes post compilation will require a larger engineering effort which I doubt we have capacity to execute on but having a use case example could make a case for it.

**Comment by christian-bromann** [comment]
Thanks, let me go ahead and close the issue then once the PR with the additional documentation is merged.

One note on the performance side, we have tested the Next.js support on a large design system and I've discovered the following observation: we created a test page where we had all components of the system being rendered in different variations, about 200 of them. This obviously had an impact on the initial page load time, not critical but also probably not ideal. We would like to gather more information about real world use cases, so if you have any information for us, please share them if you can. There are certainly optimizations we can make but given our limited resources we want to ensure they are impactful.

**Comment by christian-bromann** [comment]
A fix for this bug has been published in Stencil [`v4.21.0`](https://github.com/ionic-team/stencil/releases/tag/v4.21.0)!
